Avoid sequence cache growth

tbg opened the pull request:
While we already had a mechanism in place to eagerly garbage collect sequence
cache entries, this only affected external intents (i.e. those not colocated
with the transaction entry). On that range itself, we now clear out the
sequence cache entries on EndTransaction.

Replay behavior is somewhat affected by this: A batch which contains the whole
transaction could now replay and would execute successfully. There is an "easy"
fix: Allow invalidation of individual transactions on the read timestamp cache.
Currently, each timestamp cache entry also holds a transaction ID. The reason
for this is that a transaction should always be able to write over its own
intent, but that is normally forbidden (a WriteTooOldError). Clearing out this
memorized transaction ID would prevent those replays as well.

Fixes #4413.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4469)

<!-- Reviewable:end -->

nvanbenschoten commented:
Note that `SequenceCache` growth was diagnosed in #4322 as being a primary contributor to range split latency (~70%), so needless to say, this is a welcomed change.

nvanbenschoten commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/replica.go, line 1424 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAlRyCQ2elRcRoPpH3_:-KAlRyCRn4poiOH8QtyP:783130630) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd367d3783c3c786a3fe823ba30754dedbab96ba/storage/replica.go#L1424)):*
Are there benefits to determining this in a closure? I'm fine with it, but it doesnt seem all that Go-like. Consider changing to something like this:

```
mustUpdateSeqCache := false
if ba.Txn != nil {
    _, isEndTxn := ba.GetArg(roachpb.EndTransaction)
    mustUpdateSeqCache = !isEndTxn || err != nil
}
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 4 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/replica.go, line 1424 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAlRyCQ2elRcRoPpH3_:-KAlZXIXcwy5Mbtbuom6:-215382458) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd367d3783c3c786a3fe823ba30754dedbab96ba/storage/replica.go#L1424)):*
no, I refactored (got rid of the bool).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
What's there right now <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, although there are definitely some tests that are needed before we should push this.

---

Reviewed 1 of 2 files at r2.
Review status: 5 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-:-KAliX-h5J6cWzdaClH0:88896267)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
LGTM, but you might want to get someone who knows this code a little better than me to take a look as well.

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-:-KAll_DLbEwd6GCriqyk:148787044)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@bdarnell you probably remember this stuff. Feel free to pass it on if someone else comes to mind.

bdarnell commented:
Is the second paragraph of the commit message still accurate? The first new comment in `replica_command.go` seems to argue that replay behavior is not affected by this (and if it were, I think it would be a problem worth addressing)

Assuming even 1PC transactions are still safe from replays, LGTM.

---

Review status: 4 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/sequence_cache.go, line 204 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAmP5fKiW6QWxxMZEnA:-KAmP5fKiW6QWxxMZEnB:1727181160) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/sequence_cache.go#L204)):*
Why is the second argument nil instead of `ms`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-:-KAmPtodhpPvhR-JM73q:-1838646002)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 6 files at r1, 1 of 1 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 526 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAoNVL673Q2_zWwDn9l:-KAoNVL673Q2_zWwDn9m:-54495658) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/replica_command.go#L526)):*
this sentence is awkward to read

---

*[storage/replica_command.go, line 1043 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAoO6yPuknwM1e01dzH:-KAoO6yPuknwM1e01dzI:2117119109) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/replica_command.go#L1043)):*
is this equivalent to `if status == roachpb.COMMITTED || status == roachpb.ABORTED && !shouldPoison`?

---

*[storage/replica_command.go, line 1407 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAoOPnBQHYwxOXPi8Xg:-KAoOPnCzX05PQk6tGPC:-1888884070) ([raw file](https://github.com/cockroachdb/cockroach/blob/797d6153d2d187b9ccacd4ef77c0f22c54d20aaa/storage/replica_command.go#L1407)):*
you don't need an event before? (throughout, since otherwise it's impossible to tell if some event is slow or if there is intervening uninstrumented code)

---

*[storage/sequence_cache_test.go, line 172 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAoP-GVpSrz5NV2uFdD:-KAoP-GVpSrz5NV2uFdE:1734496998) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/sequence_cache_test.go#L172)):*
nit; prefer `else if` for the second predicate and a separate message

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 532 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAoxr0Hw9Pqon-uxXpy:-KAoxr0Hw9Pqon-uxXpz:-1888486803) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/replica_command.go#L532)):*
Do you want to put an assertion in the code where BeginTransaction is which verifies it does in fact show up with a write?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469)*
<!-- Sent from Reviewable.io -->

tbg committed:
Avoid sequence cache growth

While we already had a mechanism in place to eagerly garbage collect sequence
cache entries, this only affected external intents (i.e. those not colocated
with the transaction entry). On that range itself, we now clear out the
sequence cache entries on (successful) EndTransaction.

Replay protection is very subtle. I added more comments explaining why we
don't remove sequence cache entries on external ranges, and a TODO (soon
to be turned into an issue) about a way to get that optimization as well.

Did the beginnings of plumbing #4454, but not all the way through (sprinkled
TODOs for @nvanbenschoten instead).

tschottdorf commented:
No, the second paragraph has changed. I updated (and extended) various comments. Putting in an assertion as @spencerkimball suggested breaks a lot of (hopefully only) tests. I'm still looking into that.

---

Review status: 3 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/replica_command.go, line 532 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAoxr0Hw9Pqon-uxXpy:-KApOTwT1XMeDfr94zp_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/replica_command.go#L532)):*
Done.

---

*[storage/replica_command.go, line 1043 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAoO6yPuknwM1e01dzH:-KApQKJpCxpG0-Tt4mDs:1426891138) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/replica_command.go#L1043)):*
&& is stronger than || and the enum has nothing else, so yes, but I don't want to quiz readers on that. Code has changed anyway.

---

*[storage/replica_command.go, line 1407 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAoOPnBQHYwxOXPi8Xg:-KApQpDkkwxGIXZ2Oa6P:-279800339) ([raw file](https://github.com/cockroachdb/cockroach/blob/797d6153d2d187b9ccacd4ef77c0f22c54d20aaa/storage/replica_command.go#L1407)):*
I don't think wrapping two log events around every operations is going to increase overall happiness. I prefer reading "did X" as "did X (and possibly other slow stuff before, you'll have to bisect from the last trace event)". I originally had `Epoch` functionality in my tracer for this purpose, but it seems like nobody else is doing that sort of thing.

---

*[storage/sequence_cache.go, line 204 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAmP5fKiW6QWxxMZEnA:-KApRTgpB6Fh7SX2vMD0:1128244122) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/sequence_cache.go#L204)):*
Because bad plumbing. Fixed.

---

*[storage/sequence_cache_test.go, line 172 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAoP-GVpSrz5NV2uFdD:-KApRYNrqMNd0QOiuknD:-1632329671) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/sequence_cache_test.go#L172)):*
Going to leave as is.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-:-KApOAoy6LwZULxnKr_h:-496785973)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Added the assertion, which came with a bunch of test fallout. PTAL.

---

Review status: 2 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/sequence_cache.go, line 204 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAmP5fKiW6QWxxMZEnA:-KAqFUsanNatXa1zovuX:1342824508) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/sequence_cache.go#L204)):*
Ok, actually undid the fix. The reason is that this is only plumbing, not correct plumbing. We're not updating these stats correctly at all, but fixing it here actually subtracts contributions which were never added and trips up some range stats tests. I'm going to leave this one to @nvanbenschoten instead of "fixing" tests to report obviously wrong values (negative bytes).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-:-KAqFkikIlSPwY9ZwEXR:1141541555)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/sequence_cache.go, line 204 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-KAmP5fKiW6QWxxMZEnA:-KAqMFWh06FE0Xbgmy09:1690045519) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52cd4622022c451612d49c67af320d97560a1f3/storage/sequence_cache.go#L204)):*
SGTM, I'm rebased off of this branch right now and have the fix waiting

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469)*
<!-- Sent from Reviewable.io -->

tbg committed:
Assert txn write after BeginTransaction

As suggested by @spencerkimball. The corresponding test fallout makes up most
of the diff.

tschottdorf commented:
Still looking for LGTMs. PTAL.

bdarnell commented:
LGTM

---

Review status: 1 of 8 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4469#-:-KB5DswXL5oeBJeH2F7J:676834233)*
<!-- Sent from Reviewable.io -->
